[RPC] Refactor rref_context to not use utils::Future

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51698 [RPC] delete torch/csrc/utils/future.h
* **#51697 [RPC] Refactor rref_context to not use utils::Future**

Refactors the rest of rref_context, specifically pendingOwners map and `getOwnerRRef` to use JitFuture.

Differential Revision: [D26243268](https://our.internmc.facebook.com/intern/diff/D26243268/)

rohan-varma committed:
[RPC] Refactor rref_context to not use utils::Future

Refactors the rest of rref_context, specifically pendingOwners map and `getOwnerRRef` to use JitFuture.

Differential Revision: [D26243268](https://our.internmc.facebook.com/intern/diff/D26243268/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC] Refactor rref_context to not use utils::Future"


Refactors the rest of rref_context, specifically pendingOwners map and `getOwnerRRef` to use JitFuture.

Differential Revision: [D26243268](https://our.internmc.facebook.com/intern/diff/D26243268/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC] Refactor rref_context to not use utils::Future"


Refactors the rest of rref_context, specifically pendingOwners map and `getOwnerRRef` to use JitFuture.

Differential Revision: [D26243268](https://our.internmc.facebook.com/intern/diff/D26243268/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC] Refactor rref_context to not use utils::Future"


Refactors the rest of rref_context, specifically pendingOwners map and `getOwnerRRef` to use JitFuture.

Differential Revision: [D26243268](https://our.internmc.facebook.com/intern/diff/D26243268/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -385,21 +389,27 @@ std::shared_ptr<Future<c10::intrusive_ptr<OwnerRRef>>> RRefContext::
+    // Marks IValue Future as completed with the RRef IValue.
+    auto owner = iter->second;
+    auto rrefPtr =
+        c10::static_intrusive_pointer_cast<c10::RRefInterface>(owner);
```
`fromOwnerRRef`?

mrshenli commented on torch/csrc/distributed/rpc/rref_context.h:
```diff
@@ -105,7 +102,7 @@ class TORCH_API RRefContext {
   // used to ensure that the rref is created on the owner, otherwise throw in
   // cases where the user of this API expects this to return a completed future.
-  std::shared_ptr<Future<c10::intrusive_ptr<OwnerRRef>>> getOwnerRRef(
+  c10::intrusive_ptr<JitFuture> getOwnerRRef(
```
Since `JitFuture` does not explicitly name a type in the signature, shall we add a comment to explain what's in the JiFuture?

mrshenli commented on torch/csrc/distributed/rpc/rref_context.h:
```diff
@@ -252,10 +249,7 @@ class TORCH_API RRefContext {
-      RRefId,
-      std::shared_ptr<Future<c10::intrusive_ptr<OwnerRRef>>>,
-      RRefId::Hash>
+  std::unordered_map<RRefId, c10::intrusive_ptr<JitFuture>, RRefId::Hash>
```
ditto

mrshenli commented on torch/csrc/distributed/rpc/utils.h:
```diff
@@ -87,6 +88,18 @@ TORCH_API void populateRemoteProfiledEvents(
     const std::vector<std::vector<torch::autograd::profiler::LegacyEvent>>&
         eventLists);
 
+// Casts from c10::RRefInterface to OwnerRRef
+inline TORCH_API c10::intrusive_ptr<OwnerRRef> fromRRefInterface(
+    const c10::intrusive_ptr<c10::RRefInterface>& rrefInterface) {
+  return c10::static_intrusive_pointer_cast<OwnerRRef>(rrefInterface);
+}
+
+// Casts from OwnerRRef to c10::RRefInterface
+inline TORCH_API c10::intrusive_ptr<c10::RRefInterface> fromOwnerRRef(
+    const c10::intrusive_ptr<OwnerRRef>& ownerRRef) {
+  return c10::static_intrusive_pointer_cast<c10::RRefInterface>(ownerRRef);
+}
```
looks like all callsites of these two functions already depends on `rref_impl.h`. Shall we move these functions there to reduce the exposure and avoid soft circular dependency?

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -385,21 +389,27 @@ std::shared_ptr<Future<c10::intrusive_ptr<OwnerRRef>>> RRefContext::
     if (pendingOwnerIter == pendingOwners_.end()) {
+      // Note: The type passed into RRefType::create() does not matter here, as
+      // the future is marked as completed with the RRef of the specific type
+      // in getOrCreateOwnerRRef().
```
is there any concern if RRefType's type does not match with the real RRef value?

cc @wanchaol

rohan-varma replied:
I didn't see any failures with this presumably since we are filling it with the right type (more specific than `Any` which is the most broad) when the owner RRef is created. Although, would be good for @wanchaol to clarify this.

wanchaol replied:
it should be fine as we fill the type later when we actually create the rref, here is merely a placeholder for this rref_id i think.

rohan-varma committed:
Update on "[RPC] Refactor rref_context to not use utils::Future"


Refactors the rest of rref_context, specifically pendingOwners map and `getOwnerRRef` to use JitFuture.

Differential Revision: [D26243268](https://our.internmc.facebook.com/intern/diff/D26243268/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC] Refactor rref_context to not use utils::Future"


Refactors the rest of rref_context, specifically pendingOwners map and `getOwnerRRef` to use JitFuture.

Differential Revision: [D26243268](https://our.internmc.facebook.com/intern/diff/D26243268/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC] Refactor rref_context to not use utils::Future"


Refactors the rest of rref_context, specifically pendingOwners map and `getOwnerRRef` to use JitFuture.

Differential Revision: [D26243268](https://our.internmc.facebook.com/intern/diff/D26243268/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC] Refactor rref_context to not use utils::Future"


Refactors the rest of rref_context, specifically pendingOwners map and `getOwnerRRef` to use JitFuture.

Differential Revision: [D26243268](https://our.internmc.facebook.com/intern/diff/D26243268/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC] Refactor rref_context to not use utils::Future"


Refactors the rest of rref_context, specifically pendingOwners map and `getOwnerRRef` to use JitFuture.

Differential Revision: [D26243268](https://our.internmc.facebook.com/intern/diff/D26243268/)

[ghstack-poisoned]

wanchaol reviewed (approved):
looks good to me.

rohan-varma committed:
Update on "[RPC] Refactor rref_context to not use utils::Future"


Refactors the rest of rref_context, specifically pendingOwners map and `getOwnerRRef` to use JitFuture.

Differential Revision: [D26243268](https://our.internmc.facebook.com/intern/diff/D26243268/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[RPC] Refactor rref_context to not use utils::Future"


Refactors the rest of rref_context, specifically pendingOwners map and `getOwnerRRef` to use JitFuture.

Differential Revision: [D26243268](https://our.internmc.facebook.com/intern/diff/D26243268/)

[ghstack-poisoned]
